PlayFabSDK/source/Qos/PlayFabQosApi.cs

```diff
@@ -0,0 +1,163 @@
+﻿namespace PlayFab.QoS
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Linq;
+    using System.Threading.Tasks;
+    using EventsModels;
+    using Internal;
+    using MultiplayerModels;
+
+    public class PlayFabQosApi
+    {
+        private const int DefaultTimeoutMs = 250;
+        private readonly Dictionary<AzureRegion, string> _dataCenterMap = new Dictionary<AzureRegion, string>();
+        private readonly Action<string> _logAction;
+
+        public PlayFabQosApi(Action<string> logAction = null)
+        {
+            _logAction = logAction;
+        }
+
+        public async Task<QosResult> GetQosResultAsync(int timeoutMs = DefaultTimeoutMs)
+        {
+            QosResult result = await GetResultAsync(timeoutMs);
+            if (result.ErrorCode != 0)
+            {
+                return result;
+            }
+
+            bool eventSent = await SendResultsToPlayFab(result);
+            if (!eventSent)
+            {
+                result.ErrorCode = -3;
+            }
+
+            return result;
+        }
+
+        private async Task<QosResult> GetResultAsync(int timeoutMs)
+        {
+            var result = new QosResult();
+
+            if (!PlayFabClientAPI.IsClientLoggedIn())
+            {
+                LogQos("Client is not logged in");
+                result.ErrorCode = -1;
+                return result;
+            }
+
+            // get datacenter map (call thunderhead)
+            await GetQoSServerList();
+
+            int serverCount = _dataCenterMap.Count;
+            if (serverCount < 0)
+            {
+                result.ErrorCode = -2;
```

Comment:
Can we set these numbers to an enum that is cast to an int so that we can read specifically what went wrong?

Reply:
We can at least also LogQos, but having a stack that is understandable is also a nice thing to aim for if the logger is optional

Reply:
Enums are good too, I like enums.

Reply:
Will make the change